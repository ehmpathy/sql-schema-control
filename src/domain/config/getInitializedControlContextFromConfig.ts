/*
  1. get the config (validated and hydrated)
  2. initialize the environment based on the config
    - start the db connection
    - initialize the control environment
  3. return the language+dialect, connection, and definitions
*/
