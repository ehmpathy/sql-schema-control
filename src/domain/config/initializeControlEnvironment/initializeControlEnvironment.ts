/*
  complete any actions required to initialize control environment
    1. connect to db
    2. provision change log table
*/
