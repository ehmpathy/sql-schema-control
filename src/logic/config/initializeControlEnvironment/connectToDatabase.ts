import mysql from 'mysql2/promise';
import { DatabaseConnection, DatabaseLanguage, ControlConfig, ConnectionConfig } from '../../../types';

// create the connection to database
const connectionAdapters = {
  [DatabaseLanguage.MYSQL]: async ({ connectionConfig }: { connectionConfig: ConnectionConfig }): Promise<DatabaseConnection> => {
    const connection = await mysql.createConnection({
      host: connectionConfig.host,
      port: connectionConfig.port,
      database: connectionConfig.schema,
      user: connectionConfig.username,
      password: connectionConfig.password,
    });
    return {
      query: async ({ sql, values }: { sql: string, values?: any[] }) => {
        return connection.query(sql, values);
      },
      end: async () => connection.end(),
    };
  },
};

export const connectToDatabase = async ({ config }: { config: ControlConfig }): Promise<DatabaseConnection> => {
  // 1. get the connection adapter for the method
  const getDbConnection = connectionAdapters[config.language];

  // 2. attempt to connect
  const connection = await getDbConnection({ connectionConfig: config.connection });

  // 3. run a test query to check connection
  await connection.query({ sql: 'SHOW TABLES' }); // TODO: make this not mysql specific

  // 4. return the connection
  return connection;
};
