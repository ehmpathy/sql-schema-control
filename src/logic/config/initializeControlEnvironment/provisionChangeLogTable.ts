import { DatabaseConnection } from '../../../types';
import { readFileAsync } from '../_utils/readFileAsync';

export const provisionChangeLogTable = async ({ connection }: { connection: DatabaseConnection }) => {
  // create the table if not exists (sql consideres its existance for us)
  const createTableSql = await readFileAsync({
    filePath: `${__dirname}/../../../../schema/tables/schema_control_change_log.mysql.sql`,
  }); // TODO: make independent of language (mysql currently)
  await connection.query({ sql: createTableSql });
};
