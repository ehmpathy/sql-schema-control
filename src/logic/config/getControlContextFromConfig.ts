import { ControlContext, ResourceDefinition } from '../../types';
import { getUncontrolledResources } from '../schema/resourceDefinition/getUncontrolledResources';
import { getConfig } from './getConfig';
import { initializeControlEnvironment } from './initializeControlEnvironment';

/*
  1. get the config (validated and hydrated)
  2. initialize the environment based on the config
    - start the db connection
    - initialize the control environment
  3. return the control context
*/
export const getControlContextFromConfig = async ({
  configPath,
  strict = false,
}: {
  configPath: string;
  strict?: boolean;
}) => {
  // 1. get the config
  const config = await getConfig({ configPath });

  // 2. initialize the environment based on the config
  const { connection } = await initializeControlEnvironment({ config });

  // 3. define the definitions such that we can append to them if we want (e.g., if want to add uncontrolled)
  const definitions = [...config.definitions];

  // 4. get uncontrolled resources, if strict
  if (config.strict || strict) {
    const controlledResources = definitions.filter(
      (def) => def.constructor === ResourceDefinition,
    ) as ResourceDefinition[];
    const uncontrolledResources = await getUncontrolledResources({ connection, controlledResources });
    definitions.push(...uncontrolledResources); // append the uncontrolled resources to the definitions
  }

  // 5. return the control context
  return new ControlContext({
    language: config.language,
    dialect: config.dialect,
    connection, // note: the connection is a live dbConnection
    definitions, // note: uncontrolled resources will already have statuses
  });
};
