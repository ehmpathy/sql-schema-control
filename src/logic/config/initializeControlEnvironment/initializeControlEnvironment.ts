import { ControlConfig } from '../../../domain';
import { connectToDatabase } from './connectToDatabase';
import { provisionChangeLogTable } from './provisionChangeLogTable';

/*
  complete any actions required to initialize control environment
    1. connect to db
    2. provision change log table
*/
export const initializeControlEnvironment = async ({
  config,
}: {
  config: ControlConfig;
}) => {
  // 1. connect to db
  const connection = await connectToDatabase({ config });

  // 2. provision the change log table
  await provisionChangeLogTable({ connection });

  // 3. return the connection
  return { connection };
};
