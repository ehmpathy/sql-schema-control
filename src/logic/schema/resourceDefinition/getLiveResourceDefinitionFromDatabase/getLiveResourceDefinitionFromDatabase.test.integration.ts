import {
  getLiveResourceDefinitionFromDatabase,
  ResourceDoesNotExistError,
} from './getLiveResourceDefinitionFromDatabase';
import {
  DatabaseConnection,
  DatabaseLanguage,
  ControlConfig,
  ResourceDefinition,
  ResourceType,
} from '../../../../types';
import { promiseConfig } from './__test_assets__/connection.config';
import { initializeControlEnvironment } from '../../../config/initializeControlEnvironment';

describe('getLiveResourceDefinitionFromDatabase', () => {
  let connection: DatabaseConnection;
  beforeAll(async () => {
    const config = new ControlConfig({
      language: DatabaseLanguage.MYSQL,
      dialect: '5.7',
      connection: await promiseConfig(),
      definitions: [],
      strict: true,
    });
    ({ connection } = await initializeControlEnvironment({ config })); // ensure db is provisioned and get connection
  });
  afterAll(async () => {
    await connection.end();
  });
  describe('table', () => {
    const resource = new ResourceDefinition({
      name: 'test_table_getResourceCreateStatement',
      type: ResourceType.TABLE,
      sql: 'CREATE TABLE test_table_getResourceCreateStatement ( id BIGINT )',
    });
    it('should throw an error if table is not defined', async () => {
      await connection.query({ sql: 'DROP TABLE IF EXISTS test_table_getResourceCreateStatement' }); // ensure possible previous state does not affect test
      try {
        await getLiveResourceDefinitionFromDatabase({
          connection,
          resourceName: resource.name,
          resourceType: resource.type,
        });
        throw new Error('should not reach here');
      } catch (error) {
        expect(error.constructor).toEqual(ResourceDoesNotExistError);
      }
    });
    it('should be able to get sql of an existing table', async () => {
      await connection.query({ sql: 'DROP TABLE IF EXISTS test_table_getResourceCreateStatement' }); // ensure possible previous state does not affect test

      // apply the resource
      await connection.query({ sql: resource.sql });

      // get status
      const liveResource = await getLiveResourceDefinitionFromDatabase({
        connection,
        resourceName: resource.name,
        resourceType: resource.type,
      });
      expect(liveResource.constructor).toEqual(ResourceDefinition);
      expect(liveResource).toMatchSnapshot(); // log example of the sql
    });
  });
  describe('function', () => {
    const resource = new ResourceDefinition({
      name: 'f_function_for_testing_getResourceCreateStatement',
      type: ResourceType.FUNCTION,
      sql: `
CREATE FUNCTION f_function_for_testing_getResourceCreateStatement(
  in_message TEXT
)
RETURNS BINARY(32)
BEGIN
RETURN UNHEX(SHA(in_message)); -- some comment
END;
      `,
    });
    it('should throw an error if function is not defined', async () => {
      await connection.query({ sql: 'DROP FUNCTION IF EXISTS f_function_for_testing_getResourceCreateStatement;' }); // ensure possible previous state does not affect test
      try {
        await await getLiveResourceDefinitionFromDatabase({
          connection,
          resourceName: resource.name,
          resourceType: resource.type,
        });
        throw new Error('should not reach here');
      } catch (error) {
        expect(error.constructor).toEqual(ResourceDoesNotExistError);
      }
    });
    it('should be able to get sql of an existing function', async () => {
      await connection.query({ sql: 'DROP FUNCTION IF EXISTS f_function_for_testing_getResourceCreateStatement;' }); // ensure possible previous state does not affect test

      // apply the resource
      await connection.query({ sql: resource.sql });

      // get status
      const liveResource = await getLiveResourceDefinitionFromDatabase({
        connection,
        resourceName: resource.name,
        resourceType: resource.type,
      });
      expect(liveResource.constructor).toEqual(ResourceDefinition);
      expect(liveResource).toMatchSnapshot(); // log example of the sql
    });
  });
  describe('procedure', () => {
    const resource = new ResourceDefinition({
      name: 'upsert_something_for_getResourceCreateStatement',
      type: ResourceType.PROCEDURE,
      sql: `
CREATE PROCEDURE upsert_something_for_getResourceCreateStatement(
  IN in_value TEXT
)
BEGIN
  SELECT true;
END;
      `,
    });
    it('should throw an error if procedure is not defined', async () => {
      await connection.query({ sql: 'DROP PROCEDURE IF EXISTS upsert_something_for_getResourceCreateStatement;' }); // ensure possible previous state does not affect test
      try {
        await getLiveResourceDefinitionFromDatabase({
          connection,
          resourceName: resource.name,
          resourceType: resource.type,
        });
        throw new Error('should not reach here');
      } catch (error) {
        expect(error.constructor).toEqual(ResourceDoesNotExistError);
      }
    });
    it('should be able to get sql of an existing procedure', async () => {
      await connection.query({ sql: 'DROP PROCEDURE IF EXISTS upsert_something_for_getResourceCreateStatement;' }); // ensure possible previous state does not affect test

      // apply the resource
      await connection.query({ sql: resource.sql });

      // get status
      const liveResource = await getLiveResourceDefinitionFromDatabase({
        connection,
        resourceName: resource.name,
        resourceType: resource.type,
      });
      expect(liveResource.constructor).toEqual(ResourceDefinition);
      expect(liveResource).toMatchSnapshot(); // log example of the sql
    });
  });
});
