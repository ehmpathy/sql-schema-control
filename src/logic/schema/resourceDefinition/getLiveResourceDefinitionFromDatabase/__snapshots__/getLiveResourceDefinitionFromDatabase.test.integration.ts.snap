// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getLiveResourceDefinitionFromDatabase function should be able to get sql of an existing function 1`] = `
ResourceDefinition {
  "name": "f_function_for_testing_getResourceCreateStatement",
  "path": undefined,
  "sql": "CREATE DEFINER=\`root\`@\`%\` FUNCTION \`f_function_for_testing_getResourceCreateStatement\`(
  in_message TEXT
) RETURNS binary(32)
BEGIN
RETURN UNHEX(SHA(in_message)); -- some comment
END",
  "status": undefined,
  "type": "FUNCTION",
}
`;

exports[`getLiveResourceDefinitionFromDatabase procedure should be able to get sql of an existing procedure 1`] = `
ResourceDefinition {
  "name": "upsert_something_for_getResourceCreateStatement",
  "path": undefined,
  "sql": "CREATE DEFINER=\`root\`@\`%\` PROCEDURE \`upsert_something_for_getResourceCreateStatement\`(
  IN in_value TEXT
)
BEGIN
  SELECT true;
END",
  "status": undefined,
  "type": "PROCEDURE",
}
`;

exports[`getLiveResourceDefinitionFromDatabase table should be able to get sql of an existing table 1`] = `
ResourceDefinition {
  "name": "test_table_getResourceCreateStatement",
  "path": undefined,
  "sql": "CREATE TABLE \`test_table_getResourceCreateStatement\` (
  \`id\` bigint(20) DEFAULT NULL
) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_unicode_ci",
  "status": undefined,
  "type": "TABLE",
}
`;
